// rollup-plugin-postcss/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var rollupPluginutils = require('rollup-pluginutils');
var postcss = _interopDefault(require('postcss'));
var styleInject = _interopDefault(require('style-inject'));
var path = _interopDefault(require('path'));

function cwd(file) {
  return path.join(process.cwd(), file);
}

function index () {
  var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  var filter = rollupPluginutils.createFilter(options.include, options.exclude);
  var injectFnName = '__$styleInject';
  var extensions = options.extensions || ['.css', '.sss'];
  var getExport = options.getExport || function () {};

  return {
    intro: function intro() {
      return styleInject.toString().replace(/styleInject/, injectFnName);
    },
    transform: function transform(code, id) {
      if (!filter(id)) return null;
      if (extensions.indexOf(path.extname(id)) === -1) return null;
      var opts = {
        from: options.from ? cwd(options.from) : id,
        to: options.to ? cwd(options.to) : id,
        map: {
          inline: false,
          annotation: false
        },
        parser: options.parser
      };
      return postcss(options.plugins || []).process(code, opts).then(function (result) {
        var code = 'export default ' + injectFnName + '(' + JSON.stringify(result.css) + ',' + JSON.stringify(getExport(result.opts.from)) + ');';
        var map = options.sourceMap && result.map ? JSON.parse(result.map) : { mappings: '' };
        return {
          code: code,
          map: map
        };
      });
    }
  };
};

module.exports = index;
