// rax/lib/vdom/scheduler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var synchronousScheduler = function synchronousScheduler(job) {
  return job();
};

var requestAnimationFramePolyfill = function requestAnimationFramePolyfill(job) {
  return setTimeout(job, 16);
};
var animationScheduler = typeof requestAnimationFrame === 'undefined' ? typeof webkitRequestAnimationFrame === 'undefined' ? requestAnimationFramePolyfill : webkitRequestAnimationFrame : requestAnimationFrame;

var requestIdleCallbackPolyfill = function requestIdleCallbackPolyfill(job) {
  return setTimeout(job, 1024);
}; // 1s
var backgroundScheduler = typeof requestIdleCallback === 'undefined' ? requestIdleCallbackPolyfill : requestIdleCallback;

var currentScheduler = synchronousScheduler;

exports.default = {
  getScheduler: function getScheduler() {
    return currentScheduler;
  },
  setScheduler: function setScheduler(scheduler) {
    return currentScheduler = scheduler;
  },
  synchronousScheduler: synchronousScheduler,
  animationScheduler: animationScheduler,
  backgroundScheduler: backgroundScheduler
};
