// lazyimg/src/adapter.js
/*
    lazyimg 
    Author: _danz
*/
;

(function (win, lib) {
    require('appearjs');

    var adapter = {};
    var appearInstance;
    var runtimeFlags = {};

    var config = {
        'dataSrc': 'img-src', //指定图片地址的attribute名, 兼做lazy-class的作用
        'lazyHeight': 0, //以此高度提前触发懒加载
        'lazyWidth': 0 //以此宽度提前触发懒加载
    };

    function extendStrict(main, sub) {
        var ret = {};
        for (var k in main) {
            if (main.hasOwnProperty(k)) {
                ret[k] = sub.hasOwnProperty(k) ? sub[k] : main[k];
            }
        }
        return ret;
    }

    function preLoadImg (src, loadCallback, errorCallback) {
        var img = new Image();
        img.src = src
        img.onload = loadCallback ? loadCallback.bind(img) : null
        img.onerror = errorCallback ? errorCallback.bind(img) : null
    }

    function dispatchEvent (item, type, data) {
        var evt = document.createEvent('CustomEvent')
        evt.initCustomEvent(type, false, false, data)
        item.dispatchEvent(evt)
    }

    function applySrc(item, processedSrc, placeholderSrc) {
        if (!processedSrc) {
            return;
        }
        if (item.nodeName.toUpperCase() == 'IMG') {
            item.setAttribute('src', processedSrc);
            if (!placeholderSrc) {
                return
            }
            item.onerror = function () {
                this.onerror = null;
                this.src = placeholderSrc
            }
        } else {
            preLoadImg(processedSrc, function () {
                item.style.backgroundImage = 'url("' + processedSrc + '")';
                item.naturalWidth = this.width
                item.naturalHeight = this.height
                dispatchEvent(item, 'load')
            }, function () {
                preLoadImg(placeholderSrc, function () {
                    item.style.backgroundImage = 'url("' + placeholderSrc + '")';
                })
            })
        }
    }

    function init() {
        appearInstance = lib.appear.init({
            cls: 'imgtmp', //可选，需要遍历的元素
            once: true, //可选，是否只触发一次
            x: config.lazyWidth, //可选，容器右边距离x以内的元素加载，默认为0
            y: config.lazyHeight, //可选，容器底部距离y以内的元素加载，默认为0
            onAppear: function (evt) {
                var item = this;
                applySrc(item, item.getAttribute('i-lazy-src'), item.dataset.placeholder);
                item.removeAttribute('i-lazy-src');
            }
        });
    }

    adapter.logConfig = function () {
        console.log('lib-img Config\n', config);
    }

    adapter.fire = function () {

        if (!appearInstance) {
            init();
        }

        var label = 'i_' + Date.now() % 100000;
        var domList = document.querySelectorAll('[' + config.dataSrc + ']');

        [].forEach.call(domList, function (item) {
            if (item.dataset.lazy == 'false' && item.dataset.lazy != 'true') {
                applySrc(item, item.getAttribute(config.dataSrc), item.dataset.placeholder);
            } else {
                item.classList.add(label);
                item.setAttribute('i-lazy-src', item.getAttribute(config.dataSrc));
            }
            item.removeAttribute(config.dataSrc);
        });

        appearInstance.bind('.' + label);
        appearInstance.fire();
    }

    adapter.defaultSrc = 'data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==';

    lib.img = adapter;
    
    module.exports = adapter;

})(window, window['lib'] || (window['lib'] = {}));
